Transparent bridging for nullable values as Swift Optionals

chkn opened the pull request:
# Swift Optionals

Swift generally has no intrinsic support for arbitrary values being `null`. To express this, it provides an `Optional` type that you must use to wrap your value. In order to make this a natural experience for C# and F#, we don't expose the `Optional` type directly, but rather transparently bridge nullable values to that type. The following nullable types are supported:

- C# nullable reference types**
- `System.Nullable` (C# nullable value types)
- F# `Option` and `ValueOption` types

** C# nullable reference types are not reified in the type system--nullability is represented as attributes at the declaration site. Because of this, it's impossible for a generic type to determine if its type argument is a nullable reference type or not. This means that an accurate `SwiftType` can no longer be obtained from a static property on a generic type. Instead, I've created a new system for obtaining the `SwiftType`

# New SwiftType system

TL;DR: You should always use `SwiftType.Of` to obtain the `SwiftType` for any given managed type. There is a new optional argument where you can specify nullability. This argument is used when we reflect over the fields of a custom view to expose them to Swift. If we see a `NullableAttribute` on the field, we can now pass that information through to `SwiftType.Of` and the resulting `SwiftType` will then be bridged as an Optional value.

## Changes to Bindings

Instead of a public static property, we now use an attribute, `SwiftTypeAttribute`, to provide the `SwiftType`. The attribute has a virtual method `GetSwiftType` that is passed the `System.Type` on which the attribute is applied, as well as `SwiftType`s for any generic type arguments. By passing in the `SwiftType`s for the generic arguments, we can ensure that they take nullability into account.

In the general case, bindings can simply add a `SwiftImportAttribute` (which derives from `SwiftTypeAttribute`) with the library path from which the type is to be imported. That should be enough to handle generic and non-generic types. For opaque return types, or types with
a special mangling, you must specify the mangled name (for opaque return types, this is the mangled name for the descriptor).

This is a cleaner approach overall, and enables us to remove the hack for handling custom views from `SwiftType.Of`. Now, custom views are trivially handled by subclassing `SwiftTypeAttribute` to return a `CustomViewType`. Because of the way the attribute is inherited on subclasses, we can simply apply this internal subclass to our `View` base type, and then no attribute is needed by custom View subclasses implemented in user code.

chkn committed:
Standardize some FIXMEs

chkn committed:
Remove workaround for https://github.com/mono/mono/issues/17869

This fix has been shipping for awhile now.

chkn committed:
Add support for working with Swift optional values

Nullable reference and value types are automatically exposed
to Swift as Optional values.

chkn committed:
Remove mis-bound Color type for now

It was bound as a blittable struct but is non-POD.
Remove for now as it's pretty much unimplemented anyway.

chkn committed:
Rework SwiftType system to handle nullable reference types

- Breaks backward dep from Swift.Interop -> SwiftUI namespace
- Finishes the transparent optional bridging
- Adds new tests, all passing

A byproduct of this work is that types are bridged to Swift with
a custom attribute to get the SwiftType instead of having to declare
a public static SwiftType property. In the general case, bindings
can simply add a [SwiftImport] attribute with the library path
from which they are to be imported. That should be enough to handle
generic and non-generic types. For opaque return types, or types with
a special mangling, you must specify the mangled name (for opaque return
types, this is the mangled name for the descriptor).

CartBlanche commented on src/SwiftUI/Mac/NSHostingView.cs:
```diff
@@ -12,13 +12,13 @@ public unsafe class NSHostingView : NSView, /*INSUserInterfaceValidations,*/ INS
-			using (var handle = view.GetHandle ())
-				return Create (handle.Pointer, view.ViewType);
+			using (var handle = view.GetSwiftHandle ())
+				return Create (handle.Pointer, handle.SwiftType);
```
nit: In this case, could we not just pass `handle`, then obtain the Pointer/SwiftType inside the `Create` call?

chkn replied:
Great idea! I also took the opportunity to switch to using regular constructors instead of `Create` static methods

CartBlanche commented on XamMacSwiftUITest/AppDelegate.cs:
```diff
@@ -14,7 +14,7 @@ public class AppDelegate : NSApplicationDelegate
 		{
 			GC.Collect ();
-			return true;
+			return false;
```
Why don't we want the app to terminate properly? Same in the F# sample?

chkn replied:
The app is crashing when you close the window, meaning we have another memory management issue. Having it terminate serves no purpose and hides these types of issues

CartBlanche replied:
Oh another leak!

CartBlanche committed:
Make behaviour similar to F# sample.

CartBlanche commented on src/SwiftUI/SwiftUI/ModifiedOpacity.cs:
```diff
@@ -4,18 +4,11 @@
 
 namespace SwiftUI
 {
+    [SwiftImport (SwiftUILib.Path, "$s7SwiftUI4ViewPAAE7opacityyQrSdFQOMQ")]
```
Nice!

CartBlanche commented:
I'm really like this refactor. Removes a fair amount of duplication and I think overall simplifies things.

chkn committed:
[build] Use .NET Core toolchain to build and run SwiftUI.Tests.FSharp

The F# compiler used with the mono-based msbuild toolchain uses
FSharp.Core 4.5.2, which causes our ValueOption tests to fail.

chkn committed:
[build] Use the Release configuration consistently on CI

chkn committed:
Use regular constructors instead of Create static methods

This is more consistent with other Xamarin APIs

chkn committed:
[build] Enable link SDK for the release configuration of the test apps

This works around the bots having mismatched Xcode and Xamarin versions.

Also, don't use diag verbosity- it's almost never needed and makes the
output impossible to read.
